torchvision/datasets/hpatches.py

```diff
@@ -0,0 +1,245 @@
+            triplet networks, and sets for generic metric learning method. Default is
+            'pairs'.
+        n_samples (int, optional): Number of items to generate at each epoch. Default
+            is 1e7.
+        cache_to_ram (bool, optional): Whether to cache to all sequences to RAM to avoid
+            constant disk I/O. Default is True. Note that this needs ~10GB of free memory.
+        n_negs (int, optional): Number of negative patches returned in the case of the
+            'sets' output. Default is 16.
+    """
+    url = 'http://icvl.ee.ic.ac.uk/vbalnt/hpatches/hpatches-release.tar.gz'
+    splits_url = 'http://icvl.ee.ic.ac.uk/vbalnt/hpatches/splits.json'
+    filename = 'hpatches-release.tar.gz'
+    folder = 'hpatches-release'
+    md5 = '0ab830d37fceb2b4c86cb1cc6cc79a61'
+
+    def __init__(self, root, split='full', transform=None, download=False, output="pairs",
+                 n_samples=1e7, cache_to_ram=True, n_negs=16):
+        self.root = os.path.expanduser(root)
+        self.transform = transform
+        self.split = split
+        self.output = output
+        self.n_samples = int(n_samples)
+        self.cache_to_ram = cache_to_ram
+        self.n_negs = n_negs
+
+        if not os.path.isfile(os.path.join(self.root, 'splits.json')):
+            download_url(self.splits_url, self.root, 'splits.json', 'b08cae8889120e339f5512c10fad4d7f')
+        self.all_splits = json.load(open(os.path.join(self.root, 'splits.json')))
+        self.all_seqs = self.all_splits['full']['test']
+        self.all_seqs_all_tps = list(itertools.product(self.all_seqs, tps))
+
+        if download:
+            self.do_download()
+
+        try:
+            self.sequences = self.all_splits[self.split][splits_seqs[self.split]]
+        except:
+            raise RuntimeError('Unknown split. ' +
+                               'Valid ones are: a,b,c,illum,view,full')
+
+        if cache_to_ram:
+            manager = Manager()
+            d = manager.dict()
+            job = [Process(target=self.do_cache_sequence_to_ram, args=(d, par)) for par in self.all_seqs]
+            _ = [p.start() for p in job]
+            _ = [p.join() for p in job]
+            self.cached_sequences = d
+
+        if not self.do_check_exists():
+            raise RuntimeError('Dataset not found. ' +
+                               'Please use download=True to download it')
+
+    def __getitem__(self, index):
+        """
+        Args:
+            index (int): Index
+        Returns:
+            dict: {'left':, 'right':, 'label':} when `output` is `pairs`.
+                           OR
+            dict: {'anchor':, 'pos':, 'neg':} when `output` is `triplets`.
+                           OR
+            dict: {'pos':, 'neg':} when `output` is `sets`.
+        """
+        # fix the random seed issue with numpy and multiprocessing
+        seed = random.randrange(4294967295)
```

Comment:
@alykhantejani I actually did that due to a weird behaviour of `np.random` which makes all the workers produce the same random numbers, so each mini-batch ended up being identical when run in parallel. 

More here:
https://discuss.pytorch.org/t/why-does-numpy-random-rand-produce-the-same-values-in-different-cores/12005

Reply:
I believe this can be fixed now in master if you pass the `worker_init_fn` to `DataLoader`, and in there you specify something like
```python
def worker_init_fn(worker):
    np.random.seed(worker + 1234)
```
so can you please remove this?

Reply:
Thanks @fmassa for pointing it out. 
From the docs 
> By default, each worker will have its PyTorch seed set to base_seed + worker_id, where base_seed is a long generated by main process using its RNG. However, seeds for other libraies may be duplicated upon initializing workers (w.g., NumPy), causing each worker to return identical random numbers. (See My data loader workers return identical random numbers section in FAQ.) You may use torch.initial_seed() to access the PyTorch seed for each worker in worker_init_fn, and use it to set other seeds before data loading.

So isn't it better to use 

```python
def worker_init_fn(worker):
    np.random.seed(torch.initial_seed())
```

Reply:
Also what happens if someone forgets to call this? As I understand this the default behaviour (i..e without passing this) would be problematic right?